Content.Server/Nyanotrasen/Shipyard/Systems/ShipyardSystem.Consoles.cs

```diff
@@ -0,0 +1,240 @@
+using Content.Server.Popups;
+using Content.Server.Cargo.Systems;
+using Content.Server.Shipyard.Components;
+using Content.Server.Shuttles.Components;
+using Content.Server.Station.Systems;
+using Content.Shared.GameTicking;
+using Content.Shared.Shipyard.Events;
+using Content.Shared.Shipyard.BUI;
+using Content.Shared.Shipyard.Prototypes;
+using Content.Shared.Shipyard.Components;
+using Content.Server.Cargo.Components;
+using Robust.Server.GameObjects;
+using Robust.Shared.Audio;
+using Robust.Shared.Player;
+using Robust.Shared.Players;
+using Robust.Shared.Prototypes;
+using Content.Shared.Access.Systems;
+using Content.Shared.Access.Components;
+using Content.Shared.Shipyard;
+
+namespace Content.Server.Shipyard.Systems
+{
+    public sealed class ShipyardConsoleSystem : SharedShipyardSystem
+    {
+        [Dependency] private readonly AccessReaderSystem _accessSystem = default!;
+        [Dependency] private readonly PopupSystem _popup = default!;
+        [Dependency] private readonly UserInterfaceSystem _uiSystem = default!;
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+        [Dependency] private readonly ShipyardSystem _shipyard = default!;
+        [Dependency] private readonly StationSystem _station = default!;
+        [Dependency] private readonly CargoSystem _cargo = default!;
+
+        public void InitializeConsole()
+        {
+            SubscribeLocalEvent<ShipyardConsoleComponent, ShipyardConsolePurchaseMessage>(OnPurchaseMessage);
+            //SubscribeLocalEvent<ShipyardConsoleComponent, ShipyardConsoleSellMessage>(OnSellMessage);
+            SubscribeLocalEvent<ShipyardConsoleComponent, BoundUIOpenedEvent>(OnConsoleUIOpened);
+            //SubscribeLocalEvent<ShipyardConsoleComponent, ComponentInit>(OnInit);
+            SubscribeLocalEvent<RoundRestartCleanupEvent>(Reset);
+        }
+
+        private void OnInit(EntityUid uid, SharedShipyardConsoleComponent orderConsole, ComponentInit args)
+        {
+            //_shipyard.SetupShipyard(); ///if we have to start up the shipyard from here later
+        }
+
+        private void Reset(RoundRestartCleanupEvent ev)
+        {
+            Reset();
+        }
+
+        private void Reset()
+        {
+            //round cleanup event in case of needing OnInit;
+        }
+
+        public void OnPurchaseMessage(EntityUid uid, SharedShipyardConsoleComponent component, ShipyardConsolePurchaseMessage args)
+        {
+            if (args.Session.AttachedEntity is not { Valid: true } player)
+            {
+                return;
+            }
```

Comment:
I was not aware of this, is there a better way of ensuring the player uid is valid?

Reply:
It depends on what you're checking for. If you want an EntityUid that exists, you'd use `!Deleted(uid)`. If you want an entity that corresponds to some pattern, you'd check for a specific component with `HasComp<T>(uid)` or maybe a tag, depending on the situation. 

On further review of this particular situation, it's probably enough to check that the value is null, since I don't think AttachedEntity is ever assigned a value that isn't a valid EntityUid or null.

If you're only using the `(x is y {} z)` pattern to get a shorthand for `args.Session.AttachedEntity` as `player`, that might be okay too, but it's not one I use often, and the times I've seen it used, it's confused me as a newcomer to C#. I'm more accustomed to writing something like:
```csharp
var player = args.Session.AttachedEntity;

if (player == null)
    return;
```
This feels like it better conveys the author's intent behind what's going on here.

@Elijahrane can probably offer more insight into the particular C# idioms going on here.

Reply:
ok I looked it over and in this particular case I'm using it to explicitly cast AttachedEntity to player only if AttachedEntity.HasValue = true.  I could write it how you suggest but I would later need to do an explicit cast.

However I just reused this later on when I didnt actually need an explicit cast so I changed the other instances